### PR TITLE
fix: corrupt data using soft serial driver

### PR DESCRIPTION
**Pull request by mha1**
Fixes https://github.com/EdgeTX/edgetx/issues/3285

Summary of changes after discussion with @raphaelcoeffic:
This sets the bit start detection IRQ priority to the value used before the major serial/port API change. Tested on TX16s and 
X10express. Logic analyzer confirms great performance and spot on timing of the soft serial driver introduced with the serial/port API change.

TODO: the same or similar fix must be applied for targets Taranis and NV14 as at least external telemetry protocol LemonDSP uses soft serial. Need discussion how to go about this in a follow up PR.

### CODE REVIEW

**Comment by richardclli** [comment]
Yes, please go ahead for Taranis target as well.  I check the old driver, the priority is 0 in old driver.
NV14 do not have this config, as NV14 hardware can support hardware invert of all ext serial, so softserial is never required.

**Comment by mha1** [comment]
After discussion witg @raphaelcoeffic:
 Update for target Taranis: EXTI9_5 is a shared interrupt for Rotary Encoder and Heartbeat for X7, X7-ACCESS und TPRO radios. Changed IRQ priority accordingly

File: radio/src/targets/taranis/hal.h [hunk 1]
```diff
@@ -223,7 +223,7 @@
   // ROTARY_ENCODER_EXTI_LINE1 IRQ
   #if !defined(USE_EXTI9_5_IRQ)
     #define USE_EXTI9_5_IRQ
-    #define EXTI9_5_IRQ_Priority 5
+    #define EXTI9_5_IRQ_Priority 0
```
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/hal.h:226]: Please use some `define` to make the intent clearer. Something like this up-front:
```
#define TELEMETRY_EXTI_PRIO 0
```

And then using this everywhere you change the prio:
```
  #define EXTI9_5_IRQ_Priority TELEMETRY_EXTI_PRIO
```

**Comment by raphaelcoeffic** [comment]
As it seems, this won't work as we expect it. We cannot static priorities because of the heartbeat sync using FreeRTOS functions which MUST NOT be called from a ISR with a prio higher than the FreeRTOS SysTick ISR (see [here](https://www.freertos.org/RTOS-Cortex-M3-M4.html))

**Comment by ParkerEde** [comment]
I have tested this PR once with a QX7S ACCESS. As soon as you open a model preset where the internal ISRM is active, the radio crashes. If a model with active internal ISRM is already active, the start ends in a bootloop.

**Comment by mha1** [comment]
@ParkerEde This is the reason why the PR went back do draft. Raphael explained the root cause which makes the present solution not as universal as it should be as causes some radios like the X7access (see list above) to run into severe problems. Going back to the drawing board ...

**Comment by mha1** [comment]
@raphaelcoeffic parkerEde tested this successfully on his X7Access using the same setup he reported the crashes with: https://github.com/EdgeTX/edgetx/pull/3329#issuecomment-1475417281

**Comment by mha1** [comment]
ACCESS heartbeat and IRQ priority change hate each other. Intermittent "telemetry lost" with ACCESS telemetry.

**Comment by mha1** [comment]
Back in the game. I missed the fact that the mixer scheduler timer was initialized to exclude interrupts from update events. But generating an update event by software was the core of the idea to avoid making lethal FreeRTOS calls from within a prio 0 ISR.

Chain of events with having the shared EXTI interrupt for soft serial and heartbeat elevated to prio 0 to stop soft serial from producing corrupt data:
- EXTI detects heartbeat edge on heartbeat pin
- EXTI interrupt with prio 0 fires and starts heartbeat ISR
- heartbeat ISR resets the mixer scheduler counter as before
- heartbeat ISR generates an update event for the mixer timer scheduler
- heartbeat ISR finishes
- mixer scheduler timer fires interrupt (at prio 5) as there is a pending update event pending
- mixer scheduler timer ISR clears interrupt flag and reinitializes the timer with the appropriate mixer scheduler period
- mixer scheduler timer ISR makes FreeRTOS calls to directly do a context switch to the highest priority blocked task (mixer task)
- the blocked mixer task (waiting for the trigger) resumes and is in sync with the heartbeat

Tested on QX7access. Both Access (including telemetry) and soft serial (no data corruption) are working.

**Comment by ParkerEde** [comment]
I have now created this PR together with
#3352
#3349
#3374
#3382
tested. Both with a Horus X10S Express, and with a Taranis QX7S ACCESS. The functions ext. PPM Telem: MLINK, as well as internal ISRM with R8pro and several sensors were checked. Everything runs smoothly. Good work.

File: radio/src/targets/taranis/hal.h (hunk 1 above)
Comment (mha1) [reply radio/src/targets/taranis/hal.h]: done

**Comment by ParkerEde** [comment]
I tested this PR again with the latest MAIN+#3352 using Horus X10 Express and Taranis QX7 ACCESS. Everything runs great.

File: radio/src/targets/common/arm/stm32/heartbeat_driver.cpp [hunk 2]
```diff
@@ -42,7 +42,13 @@ static void trigger_intmodule_heartbeat()
...
 
   mixerSchedulerResetTimer();
-  mixerSchedulerISRTrigger();
+
+  // Generate a timer update event (TIM_EGR_UG) to reload the Prescaler and the repetition 
+  // counter value immediately to avoid making FreeRTOS calls within this ISR:
+  // - fires MIXER_SCHEDULER_TIMER interrupt after returning from this ISR
+  // - MIXER_SCHEDULER_TIMER_IRQHandler(void) takes care of making FreeRTOS calls
+  //   to ensure switching to highest priority task.
+  MIXER_SCHEDULER_TIMER->EGR = TIM_EGR_UG;
```
Comment (raphaelcoeffic) [inline radio/src/targets/common/arm/stm32/heartbeat_driver.cpp:51]: Can I please have a method in `mixer_scheduler_driver.cpp` that does just that please? (something like `mixerSchedulerSoftTrigger()`)

File: radio/src/targets/taranis/hal.h [hunk 3]
```diff
@@ -1962,7 +1964,7 @@
 // TELEMETRY_EXTI IRQ
 #if !defined(USE_EXTI9_5_IRQ)
   #define USE_EXTI9_5_IRQ
-  #define EXTI9_5_IRQ_Priority 5
+  #define EXTI9_5_IRQ_Priority          TELEMETRY_EXTI_PRIO
```
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/hal.h:1967]: Sorry for the mess, but I think we could have something like:
```c
#if !defined(USE_EXTI9_5_IRQ)
  #define USE_EXTI9_5_IRQ
#endif
// overwrite priority
#undef EXTI9_5_IRQ_Priority
#define EXTI9_5_IRQ_Priority          TELEMETRY_EXTI_PRIO
```

That should allow for removing the other definitions/changes in this file.

File: radio/src/targets/common/arm/stm32/heartbeat_driver.cpp (hunk 2 above)
Comment (mha1) [reply radio/src/targets/common/arm/stm32/heartbeat_driver.cpp]: done

File: radio/src/targets/taranis/hal.h (hunk 3 above)
Comment (mha1) [reply radio/src/targets/taranis/hal.h]: done

**Comment by mha1** [comment]
@pfeerick not happy with this?

**Comment by pfeerick** [comment]
More it got lost in amongst all the weeds... and keep putting it off due to other module issues... but it might as well join the party now.

**Comment by mha1** [comment]
@pfeerick Thanks you very much. I know there is more module/api related stuff but think having smaller PR's addressing specific issues is not a bad approach also for documentation and back tracking purposes.

Speaking of other module issues. I'd still be interested how #3415 is behaving for NV14. X7access, X9D+, TX16s show the expected polarities. Unfortunately I have no way of testing NV14.

**Comment by pfeerick** [comment]
Ok, poke me again in a day or two if I've not reported back on that one.

**Comment by mha1** [comment]
@pfeerick poke

**Pull request merged**